k2/python/csrc/torch/torch_util.cu

```diff
@@ -24,4 +24,53 @@ torch::DeviceType ToTorchDeviceType(DeviceType type) {
   }
 }
 
+DeviceType FromTorchDeviceType(const torch::DeviceType &type) {
+  switch (type) {
+    case torch::kCUDA:
+      return kCuda;
+    case torch::kCPU:
+      return kCpu;
+    default:
+      K2_LOG(FATAL) << "Unsupported device type: " << type
+                    << ". Only torch::kCUDA and torch::kCPU are supported";
+      return kUnk;  // unreachable code
+  }
+}
+
+template <>
+torch::Tensor ToTensor(Array1<Arc> &array) {
+  auto device_type = ToTorchDeviceType(array.Context()->GetDeviceType());
+  int32_t device_id = array.Context()->GetDeviceId();
+  auto device = torch::Device(device_type, device_id);
+  auto scalar_type = ToScalarType<int32_t>::value;
+  // an Arc has 4 members
+  torch::IntArrayRef strides = {4, 1};  // in number of elements
+  auto options = torch::device(device).dtype(scalar_type);
+
+  // NOTE: we keep a copy of `array` inside the lambda
+  // so that `torch::Tensor` always accesses valid memory.
+  return torch::from_blob(
+      array.Data(), {array.Dim(), 4}, strides, [array](void *) {}, options);
```

Comment:
I'm a little unclear what the lambda
`[array] (void *) {}`
is doing, and why it needs to capture `array` if it is doing nothing like it seems to.

Reply:
`torch::Tensor` shares the same memory with `array`.  What happens if  we try  to access the returned `tensor` after
`array` is freed?

`[array] (void *) {}` is a deleter for the `torch::Tensor`. In our case, it deletes nothing, since the memory
is managed by the `array`. But we have to ensure that the `array` outlives the returned `tensor`.

The lamda deleter captures a copy of `array` to keep it from being freed.

---

You can consider this as a trick, though I do not like tricks.

PyTorch also uses this kind of technique for its `from_numpy()` method. Refer to
https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/tensor_numpy.cpp#L178

```cpp
  Py_INCREF(obj);
  return at::from_blob(
      data_ptr,
      sizes,
      strides,
      [obj](void* data) {
        pybind11::gil_scoped_acquire gil;
        Py_DECREF(obj);
      },
      at::device(kCPU).dtype(numpy_dtype_to_aten(PyArray_TYPE(array)))
  );
```